cosmetics-web/app/forms/responsible_persons/notifications/components/bulk_ingredient_upload_form.rb

```diff
@@ -0,0 +1,170 @@
+require "csv"
+
+module ResponsiblePersons::Notifications::Components
+  class BulkIngredientUploadForm < Form
+    class IngredientCanNotBeSavedError < ArgumentError; end
+    MAX_FILE_SIZE = 15 * 1000
+
+    attribute :file
+    attribute :component
+
+    validate :file_size_validation
+    validate :file_is_csv_file_validation
+    validate :file_is_not_empty_validation
+    validate :header_missing_validation
+    validate :correct_ingredients_validation
+
+    # Original valid is reseting error messages.
+    # We have an edge case here where we are adding error messages after
+    # validation - during actual ingredient creation
+    def valid?
+      component.ingredients.delete_all
+      return false if errors.present?
+
+      super
+    rescue ActiveRecord::StatementInvalid
+      errors.add(:file, "File has incorrect characters. Please check and try again")
+      false
+    end
+
+    def save_ingredients
+      return false unless valid?
+
+      # Despite validations above, there might be rare case when ingredient can not be saved, eg:
+      # * there will be duplicated ingredient in the file
+      # * between validation and creation ingredient will be created (very rare edge case)
+      # * difference between ingredient form and model validations
+      # * internal ActiveRecord issue
+      ActiveRecord::Base.transaction do
+        @ingredients.each_with_index do |ingredient, i|
```

Comment:
Suggestion: Moving the deletion of the components existing ingredients from `save?` to `save_ingredients`.

Placed it inside the transaction to avoid a situation where the ingredient creation fails, but the component current ingredients are deleted anyways. 

```suggestion
      # Despite validations above, there might be rare case when ingredient can not be saved, eg:
      # * there will be duplicated ingredient in the file
      # * between validation and creation ingredient will be created (very rare edge case)
      # * difference between ingredient form and model validations
      # * internal ActiveRecord issue
      ActiveRecord::Base.transaction do
        component.ingredients.delete_all
        return false unless valid?
              
        @ingredients.each_with_index do |ingredient, i|
```